scio-core/src/main/scala/com/spotify/scio/hash/MutableScalableBloomFilter.scala

```diff
@@ -0,0 +1,215 @@
+
+/**
+ * A mutable, scalable wrapper around a Guava [[com.google.common.hash.BloomFilter BloomFilter]]
+ *
+ * Scalable bloom filters use a series of bloom filters, adding a new one and scaling its size by `growthRate`
+ * once the previous filter is saturated in order to maintain the desired false positive probability `fpProb`.
+ * A scalable bloom filter `contains` ("might contain") an item if any of its filters contains the item.
+ *
+ * Import `magnolify.guava.auto._` to get common instances of Guava [[com.google.common.hash.Funnel Funnel]]s.
+ *
+ * Not thread-safe.
+ */
+object MutableScalableBloomFilter {
+
+  /**
+   * The default parameter values for this implementation are based on the findings in "Scalable Bloom Filters",
+   * Almeida, Baquero, et al.: http://gsd.di.uminho.pt/members/cbm/ps/dbloom.pdf
+   *
+   * @param initialCapacity   The capacity of the first filter
+   * @param fpProb            The desired overall false positive probability
+   * @param growthRate        The growth rate of each subsequent filter added to `filters`
+   * @param tighteningRatio   The tightening ratio applied to the current `fpProb` to maintain the false positive probability over the sequence of filters
+   * @tparam T                The type of objects inserted into the filter
+   * @return  A scalable bloom filter
+   */
+  def apply[T: g.Funnel](
+    initialCapacity: Long,
+    fpProb: Double = 0.03,
+    growthRate: Int = 2,
+    tighteningRatio: Double = 0.9
+  ): MutableScalableBloomFilter[T] = MutableScalableBloomFilter(
+    fpProb,
+    initialCapacity,
+    growthRate,
+    tighteningRatio,
+    fpProb,
+    0L,
+    None,
+    Nil
+  )
+
+  def toBytes[T](sbf: MutableScalableBloomFilter[T]): Array[Byte] = {
+    // serialize each of the fields, excepting the implicit funnel
+    val baos = new ByteArrayOutputStream()
+    val dos: DataOutputStream = new DataOutputStream(baos)
+
+    dos.writeDouble(sbf.fpProb)
+    dos.writeLong(sbf.headCapacity)
+    dos.writeInt(sbf.growthRate)
+    dos.writeDouble(sbf.tighteningRatio)
+    dos.writeDouble(sbf.headFPProb)
+    dos.writeLong(sbf.headCount)
+    dos.writeInt(sbf.numFilters) // count of head + all tail filters
+    sbf.head.foreach(filter => filter.writeTo(dos))
+    sbf.tail.foreach {
+      case Left(filter) => filter.writeTo(dos)
+      case Right(ser)   => dos.write(ser.filterBytes)
+    }
+    baos.toByteArray
+  }
+
+  def fromBytes[T](
+    bytes: Array[Byte]
+  )(implicit funnel: g.Funnel[T]): MutableScalableBloomFilter[T] = {
+    val bais = new ByteArrayInputStream(bytes)
+    val dis = new DataInputStream(bais)
+
+    val fpProb = dis.readDouble()
+    val headCapacity = dis.readLong()
+    val growthRate = dis.readInt()
+    val tighteningRatio = dis.readDouble()
+    val headFPProb = dis.readDouble()
+    val headCount = dis.readLong()
+    val numFilters = dis.readInt()
+    val head = if (numFilters > 0) Some(g.BloomFilter.readFrom[T](dis, funnel)) else None
```

Comment:
There's another tiny opt we can do here, if `headCount >= headCapacity` i.e. head is already full, make it `None` and leave it as part of `tail` instead?

Reply:
This feels very edge-casey. I gave it a shot but it would introduce a bug where the next BF after append would have the same size as the previous one since we're using `None` as the flag for initial-BF-creation where the parameters don't get adjusted.

The other option would be to create the next head at deserialization time but that then requires exposing some of the internals in an unaesthetic way :/

Reply:
Yeah maybe not worth it. The chance of that happening is `1/headCapacity` and increasing slim as we add more.